core/src/apps/ethereum/sign_tx.py

```diff
@@ -208,3 +292,86 @@ def check_common_fields(msg: MsgInSignTx) -> None:
 
     if msg.chain_id == 0:
         raise DataError("Chain ID out of bounds")
+
+
+async def _handle_staking_tx_stake(
+    data_reader: BufferReader,
+    msg: MsgInSignTx,
+    network: EthereumNetworkInfo,
+    to_address: bytes,
+    maximum_fee: str,
+    fee_items: Iterable[tuple[str, str]],
+) -> None:
+    from .layout import require_confirm_stake
+
+    # stake args:
+    # - arg0: uint64, source (should be 1)
+    if data_reader.remaining_count() != constants.SC_ARGUMENT_BYTES:
+        raise DataError  # wrong number of arguments for stake (should be 1)
+    if (
+        int.from_bytes(data_reader.read_memoryview(constants.SC_ARGUMENT_BYTES), "big")
+        != 1
+    ):
+        raise DataError  # wrong value of 1st argument ('source' should be 1)
+
+    await require_confirm_stake(
+        to_address,
+        int.from_bytes(msg.value, "big"),
+        maximum_fee,
+        fee_items,
+        network,
+        bool(msg.chunkify),
+    )
+
+
+async def _handle_staking_tx_unstake(
+    data_reader: BufferReader,
+    msg: MsgInSignTx,
+    network: EthereumNetworkInfo,
+    to_address: bytes,
+    maximum_fee: str,
+    fee_items: Iterable[tuple[str, str]],
+) -> None:
+    from .layout import require_confirm_unstake
+
+    # unstake args:
+    # - arg0: uint256, value
+    # - arg1:  uint16, isAllowedInterchange (bool)
+    # - arg2: uint64, source, should be 1
+    if data_reader.remaining_count() != constants.SC_ARGUMENT_BYTES * 3:
+        raise DataError  # wrong number of arguments for unstake (should be 3)
+    value = int.from_bytes(
+        data_reader.read_memoryview(constants.SC_ARGUMENT_BYTES), "big"
+    )
+    _ = data_reader.read_memoryview(constants.SC_ARGUMENT_BYTES)  # skip arg1
+    if (
+        int.from_bytes(data_reader.read_memoryview(constants.SC_ARGUMENT_BYTES), "big")
+        != 1
+    ):
+        raise DataError  # wrong value of 3rd argument ('source' should be 1)
```

Comment:
clearer way to spell those:

```
try:
   a = r.read_memoryview(...)
   b = r.read_memoryview(...)
   # ...
   if r.remaining_count():
      raise ValueError
except (ValueError, EofError):
   raise DataError(...)
```

please change that here, you can also change it in the other instance but it's not that much better when there is just a single call to the reader.